rslib/i18n_helpers/src/garbage_collection.rs

```diff
@@ -0,0 +1,222 @@
+// Copyright: Ankitects Pty Ltd and contributors
+// License: GNU AGPL, version 3 or later; http://www.gnu.org/licenses/agpl.html
+
+use std::{collections::HashSet, fs, io::BufReader};
+
+use fluent_syntax::{ast, parser};
+use lazy_static::lazy_static;
+use regex::Regex;
+use serde_json;
+use walkdir::{DirEntry, WalkDir};
+
+use crate::serialize;
+
+/// Extract references from all Rust, Python, TS, Svelte and Designer files in
+/// the `roots`, convert them to kebab case and write them as a json to the
+/// target file.
+pub fn extract_ftl_references(roots: &[&str], target: &str) {
+    let mut refs = HashSet::new();
+    for root in roots {
+        for entry in WalkDir::new(root) {
+            let entry = entry.expect("failed to visit dir");
+            if entry.file_type().is_file() {
+                extract_references_from_file(&mut refs, &entry);
+            }
+        }
+    }
+    serde_json::to_writer(
+        fs::File::create(target).expect("failed to create file"),
+        &refs,
+    )
+    .expect("failed to write file");
+}
+
+/// Delete every entry in `ftl_root` that is not mentioned in any json in
+/// `json_root`.
+pub fn remove_unused_ftl_messages(ftl_root: &str, json_root: &str) {
+    let mut used_ftls = HashSet::new();
+    for entry in WalkDir::new(json_root) {
+        let entry = entry.expect("failed to visit dir");
+        if entry.file_type().is_file()
+            && entry
+                .file_name()
+                .to_str()
+                .expect("non-unicode filename")
+                .ends_with(".json")
```

Comment:
This is done a lot for various file endings. I'm still thinking how to add a helper, without needing to implement our own  `IntoIter` with an endless type soup.

Reply:
Somewhat alleviated by f72a51a.